### PR TITLE
Recursor: don't start as root in systemd

### PR DESCRIPTION
**Pull request by pieterlexis**
### Short description
This PR also moves the socket-dir by default to /var/run/pdns-recursor. Due to `Runtimedir` in systemf, this directory is created with the correct owenership and permissions. The init-script is amended to create this directory as well and as the recursor is started as root, all is fine :).

### Checklist

I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled this code
- [x] tested this code
- [ ] included documentation (including possible behaviour changes)
- [ ] documented the code
- [ ] added or modified regression test(s)
- [ ] added or modified unit test(s)
- [ ]  checked that this code was merged to master

### CODE REVIEW

File: builder-support/debian/recursor/ubuntu-trusty/pdns-recursor.init [hunk 1]
```diff
@@ -30,7 +30,7 @@ DAEMON=/usr/sbin/$NAME
 # Derive the socket-dir setting from /etc/powerdns/recursor.conf
 # or fall back to the default /var/run if not specified there.
 PIDDIR=$(awk -F= '/^socket-dir=/ {print $2}' /etc/powerdns/recursor.conf)
-if [ -z "$PIDDIR" ]; then PIDDIR=/var/run; fi
+if [ -z "$PIDDIR" ]; then PIDDIR=/var/run/pdns-recursor mkdir -p $PIDDIR; fi
```
Comment (phonedph1) [inline builder-support/debian/recursor/ubuntu-trusty/pdns-recursor.init]: missing ; here?

File: builder-support/specs/pdns-recursor.init [hunk 2]
```diff
@@ -18,7 +18,7 @@
 RETVAL=0
 
 PIDDIR=$(awk -F= '/^socket-dir=/ {print $2}' /etc/pdns-recursor/recursor.conf)
-if [ -z "$PIDDIR" ]; then PIDDIR=/var/run; fi
+if [ -z "$PIDDIR" ]; then PIDDIR=/var/run; mkdir -p $PIDDIR; fi
```
Comment (phonedph1) [inline builder-support/specs/pdns-recursor.init]: is this one needed given the following line?

File: pdns/unix_utility.cc [hunk 3]
```diff
@@ -181,9 +181,9 @@ int Utility::gettimeofday( struct timeval *tv, void *tz )
 
 
 // Retrieves a gid using a groupname.
-int Utility::makeGidNumeric(const string &group)
+gid_t Utility::makeGidNumeric(const string &group)
 {
-  int newgid;
+  gid_t newgid;
   if(!(newgid=atoi(group.c_str()))) {
```
Comment (omoerbeek) [inline pdns/unix_utility.cc]: Error in existing code: "0" is not considered a valid gid

File: pdns/unix_utility.cc [hunk 4]
```diff
@@ -198,9 +198,9 @@ int Utility::makeGidNumeric(const string &group)
 
 
 // Retrieves an uid using a username.
-int Utility::makeUidNumeric(const string &username)
+uid_t Utility::makeUidNumeric(const string &username)
 {
-  int newuid;
+  uid_t newuid;
   if(!(newuid=atoi(username.c_str()))) {
```
Comment (omoerbeek) [inline pdns/unix_utility.cc]: Similar problem here.

**Comment by zeha** [comment]
This removed setuid/setgid from the shipped recursor.conf, causing hosts that run packages -built- with systemd support but running on sysvinit to start the recursor as root. Is this intentional?

**Comment by pieterlexis** [comment]
> This removed setuid/setgid from the shipped recursor.conf, causing hosts that run packages -built- with systemd support but running on sysvinit to start the recursor as root. Is this intentional?

I totally forgot we shipped with an init-script on systemd-enabled Debian, will PR a fix!

**Pull request merged**